zebra-state/src/service/non_finalized_state/tests/prop.rs

```diff
@@ -42,8 +43,8 @@ fn forked_equals_pushed() -> Result<()> {
             // use `fork_at_count` as the fork tip
             let fork_tip_hash = chain[fork_at_count - 1].hash;
 
-            let mut full_chain = Chain::new(network, Default::default(), Default::default(), finalized_tree.clone());
-            let mut partial_chain = Chain::new(network, Default::default(), Default::default(), finalized_tree.clone());
+            let mut full_chain = Chain::new(network, Default::default(), Default::default(), finalized_tree.clone(), ValueBalance::zero());
```

Comment:
Starting with an empty value balance in some tests might not be a good idea, if the generated block ends up with a negative value balance then the constrain will make the test fail. If we start with a value balance with max amount instead of all zeros then we will fail with any positive value balance generated by the block. 

I tried a method `half_full()` that populates the value pools with half of the `MAX_AMOUNT` however it didnt helped much in the prop tests.

Reply:
It looks like your code should work, and it's the tests that have bugs.
Getting the tests passing might be a bit complicated.

There are a few different bugs in the tests, so we might need to do different things for different tests.

Which fixes did you want to do?
Did you want me to help with any of them?

#### Spending missing UTXOs

Some tests commit blocks that spend missing UTXOs, so we could try to:
* rewrite those tests, or
* provide fake UTXOs, or
* calculate the block value balance change before we call `Chain::push`, and provide a `zero` value balance for the block.

Here is one of the failing tests and its error:
```
---- service::non_finalized_state::tests::prop::finalized_equals_pushed stdout ----
Message:  provided Outputs (length 3) don't have spent OutPoint ...
```

#### Removing missing blocks

Other tests try to remove blocks they never added to the chain, so we could try to:
* rewrite those tests, or
* set all the values in those blocks to zero.

These are the tests that fail when they start with a zero value balance.
(Other tests work fine with zero value balances.)

Here's how those tests generate blocks:
https://github.com/ZcashFoundation/zebra/blob/5c5abf61719b73af46d54473741c482b585b48fc/zebra-state/src/tests.rs#L25

Here is one of the failing tests and its error:
```
---- service::non_finalized_state::tests::vectors::shorter_chain_can_be_best_chain stdout ----
Message:  The reverse operation will leave the pools in a previously valid state: Transparent(Constraint ...
```

#### Not sure

I'm not sure why this test is failing, I'll need to add more logging to find out.

```
---- service::non_finalized_state::tests::prop::rejection_restores_internal_state stdout ----
Message:  Test failed: error adding value balances to the chain value pool: Orchard(Constraint ...
```

Reply:
So, i added a `fake_populated_pool()` method to `ValueBalance` to return a pool half full. Also added  `set_current_value_pool()` to `FinalizedState` to modify the value pool in the database in https://github.com/oxarbitrage/zebra/pull/130/commits/bc6775cdf77e6af6d6d71635a89ac26f2ebe9027 , also fixed a few tests.

Then i fixed the others that were having constrain issues in https://github.com/oxarbitrage/zebra/pull/130/commits/7db217483250efe6ce15b86ab4decada73724d63

This 2 commits can became one stand-alone pull request into zebra.

That leave us with 5 failing tests that are for other reasons:

```
service::non_finalized_state::tests::prop::finalized_equals_pushed
service::non_finalized_state::tests::prop::forked_equals_pushed
service::non_finalized_state::tests::vectors::construct_many
service::non_finalized_state::tests::vectors::construct_single
service::non_finalized_state::tests::vectors::ord_matches_work
```

* `finalized_equals_pushed` and `forked_equals_pushed` also had constrain issues that we resolved in the above commit, however they also have spent input problems.

Reply:
> service::non_finalized_state::tests::prop::finalized_equals_pushed
> service::non_finalized_state::tests::prop::forked_equals_pushed
> service::non_finalized_state::tests::vectors::construct_many
> service::non_finalized_state::tests::vectors::construct_single
> service::non_finalized_state::tests::vectors::ord_matches_work

I think these are mostly `push` tests, so I'm going to try writing this fix today:
* calculate the block value balance change before we call `Chain::push`, and provide a `zero` value balance for the block

Some time we might want to rewrite these tests to be less fragile, but we don't have to do it any time soon.